src/services/inference/module/module_js.ml

```diff
@@ -93,6 +93,69 @@ let module_name_candidates ~options =
       in
       List.rev (name :: List.fold_left map_name [] mappers))
 
+let types_versions_package_candidates reader =
+  Module_hashtables.memoize_with_types_versions_candidates_cache ~f:(fun package_filename ->
+      let package =
+        match Module_heaps.Reader_dispatcher.get_package ~reader package_filename with
+        | Some (Ok package) -> package
+        | _ -> Package_json.empty
+      in
+      let types_versions = Package_json.flow_types_versions package in
+      let version_path =
+        SMap.find_first_opt
+          (fun k ->
+            try Semver.satisfies ~include_prereleases:true k Flow_version.version
+            with Semver.Parse_error _ -> false)
+          types_versions
+      in
+      version_path)
+
+let types_versions_candidates ~reader dir rel_path dirname =
```

Comment:
This function and the associated code in `node_paths_part.ml` really make me nervous. I get worried whenever I see code that manipulates paths directly. There are a lot of opportunities to make mistakes, both for the original author and future maintainers.

I think it would help if I understood the motivation for this code better. What is happening here that couldn't be satisfied by reusing existing code? Fundamentally, it seems to me like we are looking in the `package.json` already for `main` to figure out what to resolve the module to, why can't we augment that logic to look at the `typesVersions` fields as well?

I have some more specific comments about this code, but I'd like to understand the context better before I dig deeper.

Reply:
Is it possible to write unit tests for this? I didn't found any runner in repository, but it has some OUnit tests

Reply:
We do have some unit tests, but they are only set up to run with Buck so we haven't open sourced them. @mroch, do you know what would be involved in setting up our unit tests to run with `make`?